gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/Message.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.pubsub;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.ImmutableMap;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Timestamp;
+import com.google.pubsub.v1.PubsubMessage;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Pub/Sub message.
+ */
+public class Message implements Serializable {
```

Comment:
I don't think that exposing protobuf classes as `ByteString` would be a good design choice for our API.

> For this particular class, I'm also worried about unnecessary copy of the data when doing conversions from ByteString to byte[] and back.

This is a good point. To save some unnecessary copies we could change our `payload()` getter to return a `ByteBuffer` instead of `byte[]`(`ByteString.asReadOnlyByteBuffer()` does not do a copy). And we should also provide `payload(...)` setters that take `InputStream` to avoid users to first read the stream into bytes and then copying them with our current `payload(byte[])`.